server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java

```diff
@@ -4947,6 +4974,66 @@ public void testKeepMinRetainedSeqNoByMergePolicy() throws IOException {
+            engine.refresh("test");
+            Map<Long, Long> updates = new HashMap<>();
+            try (Searcher engineSearcher = engine.acquireSearcher("test", Engine.SearcherScope.INTERNAL)) {
+                DirectoryReader reader = Lucene.wrapAllDocsLive(engineSearcher.getDirectoryReader());
+                for (LeafReaderContext leaf : reader.leaves()) {
+                    NumericDocValues seqNoDV = leaf.reader().getNumericDocValues(SeqNoFieldMapper.NAME);
+                    NumericDocValues updatedByDV = leaf.reader().getNumericDocValues(SeqNoFieldMapper.UPDATED_BY_SEQNO_NAME);
+                    DocIdSetIterator iterator = new IndexSearcher(reader).createWeight(new MatchAllDocsQuery(), false, 1.0f)
+                        .scorer(leaf).iterator();
+                    int doc;
+                    while((doc = iterator.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+                        assertThat(seqNoDV.advanceExact(doc), equalTo(true));
+                        if (updatedByDV.advanceExact(doc)) {
+                            assertThat(updatedByDV.longValue(), greaterThanOrEqualTo(0L));
+                            updates.put(seqNoDV.longValue(), updatedByDV.longValue());
+                        }
+                    }
+                }
+            }
+            return updates;
+        };
+        Settings.Builder settings = Settings.builder()
+            .put(defaultSettings.getSettings())
+            .put(IndexSettings.INDEX_SOFT_DELETES_SETTING.getKey(), true);
+        final IndexMetaData indexMetaData = IndexMetaData.builder(defaultSettings.getIndexMetaData()).settings(settings).build();
+        final IndexSettings indexSettings = IndexSettingsModule.newIndexSettings(indexMetaData);
+        // On replica
+        AtomicLong globalCheckpoint = new AtomicLong(SequenceNumbers.NO_OPS_PERFORMED);
+        try (Store store = createStore();
+             InternalEngine engine = createEngine(config(indexSettings, store, createTempDir(),
+                 newMergePolicy(), null, null, globalCheckpoint::get))) {
+            engine.index(replicaIndexForDoc(createParsedDoc("1", null), 1, 0, false));
+            engine.index(replicaIndexForDoc(createParsedDoc("1", null), 2, 1, false));
+            // IndexWriter allows softUpdateDocuments soft-deleted docs again until it's refreshed.
+            engine.refresh("test");
+            engine.delete(replicaDeleteForDoc("1", 4, 3, threadPool.relativeTimeInMillis()));
+            engine.index(replicaIndexForDoc(createParsedDoc("1", null), 3, 2, false));
+            engine.index(replicaIndexForDoc(createParsedDoc("1", null), 5, 4, false));
+            Map<Long, Long> seqNos = readUpdatedBySeqNos.apply(engine);
+            assertThat(seqNos, hasEntry(0L, 1L)); // 0 -> 1
+            assertThat(seqNos, hasEntry(1L, 4L)); // 1 -> 3 -> 4
+            assertThat(seqNos, hasEntry(2L, 3L)); // 2 -> 3 (stale)
```

Comment:
+1 . Can we also have stale delete operations ?

Reply:
why don't we have an entry for term seq3? also check that the map size is 3?

Reply:
done